File: in_a_Weekend/python/hitable.py
import math

from vec3 import Vec3

class HitRecord(object):

    def __init__(self):
        self.t = None
        self.p = None
        self.normal = None
        self.material = None

class Hitable(object):

    def hit(self, r, t_min, t_max, rec):
        raise NotImplementedError

class HitableList(Hitable):

    def __init__(self, l=None):
        if l is None: l = []
        self.list = l

    def hit(self, r, t_min, t_max):
        res = False
        closest_so_far = t_max
        for hitable in self.list:
            h = hitable.hit(r, t_min, closest_so_far)
            if not h: continue
            closest_so_far = h.t
            res = h
        return res

class Sphere(Hitable):

    def __init__(self, cen=None, r=0.0, material=None):
        if cen is None: cen = Vec3()
        self.center = cen
        self.radius = r
        self.material = material

    def hit(self, r, t_min, t_max):
        def updateRec(t):
            if not (t_min < t < t_max):
                return False
            res = HitRecord()
            res.t = t
            res.p = r.pointAt(t)
            res.normal = (res.p - self.center) / self.radius
            res.material = self.material
            return res

        oc = r.origin - self.center
        a = r.direction.dot(r.direction)
        b = oc.dot(r.direction)
        c = oc.dot(oc) - self.radius*self.radius
        discriminant = b*b - a*c
        if discriminant > 0:
            res = updateRec((-b - math.sqrt(discriminant))/a)
            if not res:
                res = updateRec((-b + math.sqrt(discriminant))/a)
            if res: return res

        return False

File: in_a_Weekend/rust/rti1w/src/hitable.rs
use vec3::Vec3;
use ray::Ray;

pub struct HitRecord {
    pub t: f32,
    pub p: Vec3,
    pub normal: Vec3
}

/*************************************************************************************************/
pub trait Hitable {
    fn hit(&self, r: &Ray, t_min: f32, t_max: f32) -> Option<HitRecord>;
}

/*************************************************************************************************/
pub struct HitableList {
    pub list: Vec<Box<Hitable>>
}

impl HitableList {
    pub fn new(list: Vec<Box<Hitable>>) -> Self {
        HitableList { list }
    }
}

impl Hitable for HitableList {
    fn hit(&self, r: &Ray, t_min: f32, t_max: f32) -> Option<HitRecord> {
        let mut best: Option<HitRecord> = None;
        for child in self.list.iter() {
            if let Some(hit) = child.hit(r, t_min, t_max) {
                if best.is_none() || hit.t < best.as_ref()?.t { best = Some(hit) }
            }
        }
        best
    }
}

/*************************************************************************************************/
pub struct Sphere {
    pub center: Vec3,
    pub radius: f32
}

impl Sphere {
    pub fn new(center: Vec3, radius: f32) -> Self {
        Sphere { center, radius }
    }
}

impl Hitable for Sphere {
    fn hit(&self, r: &Ray, t_min: f32, t_max: f32) -> Option<HitRecord> {
        let oc = r.origin - self.center;
        let a = r.direction.dot(r.direction);
        let b = oc.dot(r.direction);
        let c = oc.dot(oc) - self.radius*self.radius;
        let discriminant = b*b - a*c;
        let hit = |t: f32| -> Option<HitRecord> {
            if t <= t_min || t >= t_max { return None }
            let p = r.point_at(t);
            return Some(HitRecord { t, p, normal: (p - self.center) / self.radius })
        };
        if discriminant > 0. {
            let h = hit((-b - discriminant.sqrt())/a);
            if h.is_some() { return h }
            let h = hit((-b + discriminant.sqrt())/a);
            if h.is_some() { return h }
        }
        None
    }
}
